r/R/dplyr.R

```diff
@@ -423,26 +493,114 @@ ungroup.arrow_dplyr_query <- function(x, ...) {
 }
 ungroup.Dataset <- ungroup.ArrowTabular <- force
 
-mutate.arrow_dplyr_query <- function(.data, ...) {
+mutate.arrow_dplyr_query <- function(.data,
+                                     ...,
+                                     .keep = c("all", "used", "unused", "none"),
+                                     .before = NULL,
+                                     .after = NULL) {
+  call <- match.call()
+  exprs <- quos(...)
+  if (length(exprs) == 0) {
+    # Nothing to do
+    return(.data)
+  }
+
   .data <- arrow_dplyr_query(.data)
   if (query_on_dataset(.data)) {
     not_implemented_for_dataset("mutate()")
   }
-  # TODO: see if we can defer evaluating the expressions and not collect here.
-  # It's different from filters (as currently implemented) because the basic
-  # vector transformation functions aren't yet implemented in Arrow C++.
-  dplyr::mutate(dplyr::collect(.data), ...)
+
+  .keep <- match.arg(.keep)
+  .before <- enquo(.before)
+  .after <- enquo(.after)
+  # Restrict the cases we support for now
+  if (!quo_is_null(.before) || !quo_is_null(.after)) {
+    # TODO(ARROW-11701)
+    return(abandon_ship(call, .data, '.before and .after arguments are not supported in Arrow'))
+  } else if (length(group_vars(.data)) > 0) {
+    # mutate() on a grouped dataset does calculations within groups
+    # This doesn't matter on scalar ops (arithmetic etc.) but it does
+    # for things with aggregations (e.g. subtracting the mean)
+    return(abandon_ship(call, .data, 'mutate() on grouped data not supported in Arrow'))
+  }
+
+  # Check for unnamed expressions and fix if any
+  unnamed <- !nzchar(names(exprs))
+  # Deparse and take the first element in case they're long expressions
+  names(exprs)[unnamed] <- map_chr(exprs[unnamed], ~deparse(.)[1])
```

Comment:
I think rlang or tibble has a way of shortening long expressions, and we should use that. I'm trying to find it...

Reply:
Ah yes, it's `rlang::as_label()`

Reply:
After doing the name shortening, you should check for duplicate names and repair them, e.g. something like this
```r
dup_names <- duplicated(names(exprs)[unnamed]) & !duplicated(exprs[unnamed])
if (any(dup_names)) {
  names(exprs)[unnamed] <- vctrs::vec_as_names(names(exprs)[unnamed], repair = "unique")
}
```

Reply:
Here's an example showing why we need that:
```r
iris %>% 
  record_batch() %>% 
  transmute(
    Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length,
    Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Length + Sepal.Width,
    Width = Sepal.Width,
    Width = Petal.Width
  )
```
That should return three columns with unique names.

Reply:
Not the most compelling user story there ;) but I'll see what I can do

Reply:
Pushed https://github.com/apache/arrow/pull/9521/commits/020dad00d5e878cdef193d07577f02ddf3c2a8cf to better handle these. We don't want to dedupe the names in the expression because reusing the same name means to overwrite--this is what dplyr does (and the tests confirm we do the same).

Reply:
Ah right, ok—but we should dedupe any names that became duplicates only after we applied `as_label()`

Reply:
Wait, since this renaming only applies to columns that did not have names to begin with, we _should_ dedupe them. I think my code in the comment above (https://github.com/apache/arrow/pull/9521#discussion_r582298264) should do it right.

Reply:
I tried this, and I found that dplyr also gave two columns, not three, because it also calls `as_label`

Reply:
Ok, fair enough 👍